### ISSUE TITLE
_Fault() will be called after user thread finished if PMP_STACK_GUARD=y on RISC-V platform

### ISSUE THREAD

**Issue by atshinbo-etrinity** [zephyrproject-rtos/zephyr#39924]
**Describe the bug**
When hello_world_user is built with 
CONFIG_PMP_STACK_GUARD=y
CONFIG_HW_STACK_PROTECTION=y,
after user mode thread finished, _Fault() has been called.

**To Reproduce**
1. Modify samples/userspace/hello_world_user/src/main.c so that the link result does not exceed the RAM capacity.
```
< #define USER_STACKSIZE	2048

> #define USER_STACKSIZE	1024
```

2. Modify zephyr/include/arch/riscv/common/linker.ld as following,
(This is workaround for another issue. #39922)
```
< __rom_region_size = __rom_region_end - __rom_region_start;

> __rom_region_size_orig = __rom_region_end - __rom_region_start;
> __rom_region_size = (1 << LOG2CEIL(__rom_region_size_orig));
```

3. Add 2 lines to hello_world_user/prj.conf
```
CONFIG_PMP_STACK_GUARD=y
CONFIG_HW_STACK_PROTECTION=y
```

4. west build -b hifive1_revb samples/userspace/hello_world_user
5. cd build/zephyr
6. $ZEPHYR_SDK_INSTALL_DIR/riscv64-zephyr-elf/bin/riscv64-zephyr-elf-objcopy -O ihex zephyr.elf zephyr.hex
7. Write zephyr.hex to the Hifive1 revB board and run.

_Fault() will be called. 
However, there is no UART message because the fault will be happened recursively.
(Please see "Investigation" below)

**Investigation**

(1) Direct cause of the fault
```
zephyr/arch/riscv/core/isr.S
 926         /* Check the thread mode */
 927         WAS_NOT_USER(t2, t4)
 928         bnez t2, kernel_swap
 929 
 930         /* Switch to user stack */
 931         RV_OP_LOADREG sp, _thread_offset_to_user_sp(t1)
```
The sp register will be set to 0 in the skip_callee_saved_reg routine at line 931 of above code.

To confirm that the sp register is set to 0, 
I made the following modification temporarily and confirmed _Fault is called (t2 == 0).
```
zephyr/arch/riscv/core/isr.S:931
<   /* Switch to user stack */
<   RV_OP_LOADREG sp, _thread_offset_to_user_sp(t1)

>   /* Switch to user stack */
>   RV_OP_LOADREG t2, _thread_offset_to_user_sp(t1)
>   beqz t2, _Fault
>   mv sp, t2
```

(2) Cause of the above (1)

Above (1) is happen after user thread function was finished and thread_abort system call was called from z_thread_entry.
The path of returning from thread_abort system call is suspicious. However I don't know the detailed cause of the problem.

**Environment (please complete the following information):**
 - OS: Ubuntu 18.04.5 LTS
 - Toolchain: zephyr-sdk-0.13.1
 - Version: v2.7.99
 - Commit: f6e829a68b1ae7b127bb847d0947df0cc9f708a2
 - Target: hifive1_revb

**Comment by atshinbo-etrinity** [comment]
I have confirmed that this issue has been fixed at latest commit of main branch (e8df8a579033be238beb774e4909d637da649f3a).

It seems the issue has been fixed by following changes:
> zephyr/arch/riscv/core/isr.S 5fc5beabe2e9dd9e815f2bf13c9708ceabd858dc arch: riscv: pmp: fix IRQ handling of PMP stack guard

In my investigation, the cause of this issue is :
- "main" thread uses z_interrupt_stacks instead of z_main_stack when CONFIG_PMP_STACK_GUARD=y.
- k_thread_create() will save the context of main thread to z_interrupt_stacks. After that, when an interrupt happens, this area will be overitten.
- After "user" thread finished, the context of main thread which was corrupted is restored and an fault happen.

The latest commit fixed this behavior. "main" thread uses z_main_stack correctly even if CONFIG_PMP_STACK_GUARD=y.

**Comment by cwshu** [comment]
@atshinbo-etrinity 
Thanks for you investigation. I found same bug in QEMU, so I fixed this issue in the 2nd part of commit 5fc5bea.

`no_reschedule` will change the stack at the end of ISR, but `no_reschedule_resched` won't.
The back-to-back interrupt checking is in the `reschedule` so that it should goto `no_reschedule_resched` instead of `no_reschedule`.

p.s.
- `no_reschedule` use IRQ stack so it change stack at the end of ISR
- `reschedule` use task stack (it does the context switch) so it doesn't change stack at the end of ISR.
